File: AdSwitcher-Android/adcolony-adapter/src/main/java/net/adswitcher/adapter/adcolony/AdColonyAdapter.java
package net.adswitcher.adapter.adcolony;

import android.app.Activity;
import android.os.Handler;
import android.util.Log;

import com.adcolony.sdk.*;

import net.adswitcher.adapter.InterstitialAdAdapter;
import net.adswitcher.adapter.InterstitialAdListener;

import java.util.Map;

/**
 * Created by tkyaji on 2016/08/12.
 */
public class AdColonyAdapter implements InterstitialAdAdapter {

    private static final String TAG = "AdColonyAdapter";

    private String zoneId;
    private InterstitialAdListener interstitialAdListener;
    AdColonyInterstitial adColonyInterstitial;
    private Activity activity;
    private boolean isLoading;
    private boolean isLoaded;


    @Override
    public void interstitialAdInitialize(Activity activity, InterstitialAdListener interstitialAdListener, Map<String, String> parameters, boolean testMode) {
        this.interstitialAdListener = interstitialAdListener;
        this.activity = activity;

        String appId = parameters.get("app_id");
        this.zoneId = parameters.get("zone_id");

        Log.d(TAG, "videoAdInitialize : app_id=" + appId + ", zone_id=" + this.zoneId);
        AdColony.configure(activity, appId, this.zoneId);
    }

    @Override
    public void interstitialAdLoad() {
        Log.d(TAG, "interstitialAdLoad");

        this.adColonyInterstitial = null;
        this.isLoading = true;
        this.isLoaded = false;

        AdColony.requestInterstitial(this.zoneId, this.getListener());

        this.adLoad(1);
    }

    @Override
    public void interstitialAdShow() {
        Log.d(TAG, "videoAdShow");
        if (this.isLoaded && this.adColonyInterstitial.show()) {
        } else {
            this.interstitialAdListener.interstitialAdClosed(this, false, false);
        }
    }

    private void adLoad(final int count) {
        Log.d(TAG, "adLoad : count=" + count);

        (new Handler()).postDelayed(new Runnable() {
            @Override
            public void run() {
                if (!AdColonyAdapter.this.isLoading) {
                    return;

                } else if (count == 5) {
                    AdColonyAdapter.this.isLoading = false;
                    Log.d(TAG, "load timeout");
                    AdColonyAdapter.this.interstitialAdListener.interstitialAdLoaded(AdColonyAdapter.this, false);

                } else {
                    AdColonyAdapter.this.adLoad(count + 1);
                }
            }
        }, 1000);
    }

    private AdColonyInterstitialListener getListener() {
        return new AdColonyInterstitialListener() {
            @Override
            public void onRequestFilled(AdColonyInterstitial adColonyInterstitial) {
                Log.d(TAG, "onRequestFilled");
                AdColonyAdapter.this.adColonyInterstitial = adColonyInterstitial;
                AdColonyAdapter.this.isLoading = false;
                AdColonyAdapter.this.isLoaded = true;
                AdColonyAdapter.this.interstitialAdListener.interstitialAdLoaded(AdColonyAdapter.this, true);
            }

            @Override
            public void onClicked(AdColonyInterstitial ad) {
                Log.d(TAG, "onClicked");
                AdColonyAdapter.this.interstitialAdListener.interstitialAdClicked(AdColonyAdapter.this);
            }

            @Override
            public void onClosed(AdColonyInterstitial ad) {
                Log.d(TAG, "onClosed");
                AdColonyAdapter.this.interstitialAdListener.interstitialAdClosed(AdColonyAdapter.this, true, false);
            }

            @Override
            public void onOpened(AdColonyInterstitial ad) {
                Log.d(TAG, "onOpened");
                AdColonyAdapter.this.interstitialAdListener.interstitialAdShown(AdColonyAdapter.this);
            }
        };
    }
}

File: AdSwitcher-iOS/AdSwitcher/Adapters/AdColonyAdapter/AdColonyAdapter.m
//
//  AdColonyAdapter.m
//  AdColonyAdapter
//
//  Created by tkyaji on 2016/08/09.
//  Copyright © 2016年 adwitcher. All rights reserved.
//

#import "AdColonyAdapter.h"

@implementation AdColonyAdapter {
    NSString *_zoneId;
    UIViewController *_appViewController;
    AdColonyInterstitial *_adColonyInterstitial;
    
    BOOL _isInitialized;
    BOOL _isLoading;
    BOOL _isLoaded;
}

@synthesize interstitialAdDelegate;

# pragma - InterstitialAdAdapter

- (void)interstitialAdInitialize:(UIViewController *)viewController parameters:(NSDictionary<NSString *,NSString *> *)parameters testMode:(BOOL)testMode {
    
    NSString *appId = [parameters objectForKey:@"app_id"];
    _zoneId = [parameters objectForKey:@"zone_id"];
    
    _DLOG(@"app_id:%@, zone_id:%@", appId, _zoneId);
    
    [AdColony configureWithAppID:appId zoneIDs:@[_zoneId] options:nil completion:^(NSArray<AdColonyZone *> *zones) {
        _DLOG("%@", appId);
        _isInitialized = YES;
    }];
}

- (void)interstitialAdLoad {
    if (_isInitialized) {
        [self requestLoad];
        
    } else {
        dispatch_time_t time = dispatch_time(DISPATCH_TIME_NOW, 1 * NSEC_PER_SEC);
        dispatch_after(time, dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
            [self requestLoad];
        });
    }
}

- (void)requestLoad {
    _adColonyInterstitial = nil;
    _isLoading = YES;
    _isLoaded = NO;
    
    [AdColony requestInterstitialInZone:_zoneId options:nil success:^(AdColonyInterstitial *ad) {
        _DLOG("success");
        _adColonyInterstitial = ad;
        _isLoading = NO;
        _isLoaded = YES;
        [self.interstitialAdDelegate interstitialAdLoaded:self result:YES];
        
        [ad setOpen:^{
            _DLOG("open");
            [self.interstitialAdDelegate interstitialAdShown:self];
        }];
        [ad setClick:^{
            _DLOG("click");
            [self.interstitialAdDelegate interstitialAdClicked:self];
        }];
        [ad setClose:^{
            _DLOG("close");
            [self.interstitialAdDelegate interstitialAdClosed:self result:YES isSkipped:NO];
        }];
        
    } failure:^(AdColonyAdRequestError *error) {
        _DLOG("failure: %@", error);
        [self.interstitialAdDelegate interstitialAdLoaded:self result:NO];
        _isLoading = NO;
    }];
    
    [self adLoad:1];
}

- (void)interstitialAdShow {
    if (_isLoaded && [_adColonyInterstitial showWithPresentingViewController:_appViewController]) {
        _DLOG("show:1");
    } else {
        _DLOG("show:0");
        [self.interstitialAdDelegate interstitialAdClosed:self result:NO isSkipped:NO];
    }
}


#pragma - Private methods

- (void)adLoad:(int)count {
    _DLOG("count=%d", count);
    
    dispatch_time_t time = dispatch_time(DISPATCH_TIME_NOW, 1 * NSEC_PER_SEC);
    dispatch_after(time, dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        if (!_isLoading) {
            return;
            
        } else if (count == 5) {
            _isLoading = NO;
            _DLOG("load timeout");
            [self.interstitialAdDelegate interstitialAdLoaded:self result:NO];
        } else {
            [self adLoad:count + 1];
        }
    });
}


@end
